shared-module/bitmaptools/__init__.c

```diff
@@ -252,6 +253,95 @@ void common_hal_bitmaptools_fill_region(displayio_bitmap_t *destination,
     }
 }
 
+void common_hal_bitmaptools_boundary_fill(displayio_bitmap_t *destination,
+    int16_t x, int16_t y,
+    uint32_t value, uint32_t background_value) {
+
+    // the list of points that we'll check
+    mp_obj_t fill_area = mp_obj_new_list(0, NULL);
+
+    // first point is the one user passed in
+    mp_obj_t point[] = { mp_obj_new_int(x), mp_obj_new_int(y) };
+    mp_obj_list_append(
+        fill_area,
+        mp_obj_new_tuple(2, point)
+        );
+
+    mp_obj_t *fill_points;
+    size_t list_length = 0;
+    mp_obj_list_get(fill_area, &list_length, &fill_points);
+
+    mp_obj_t current_point;
+    uint32_t current_point_color_value;
+
+    size_t tuple_len = 0;
+    mp_obj_t *tuple_items;
+
+    // while there are still points to check
+    while (list_length > 0) {
+        mp_obj_list_get(fill_area, &list_length, &fill_points);
+        current_point = mp_obj_list_pop(fill_area, 0);
+        mp_obj_tuple_get(current_point, &tuple_len, &tuple_items);
+        current_point_color_value = common_hal_displayio_bitmap_get_pixel(
+            destination,
+            mp_obj_get_int(tuple_items[0]),
+            mp_obj_get_int(tuple_items[1]));
+
+        // if the current point is not background color ignore it
+        if (current_point_color_value != background_value) {
+            mp_obj_list_get(fill_area, &list_length, &fill_points);
+            continue;
+        }
+
+        // fill the current point with fill color
+        displayio_bitmap_write_pixel(
+            destination,
+            mp_obj_get_int(tuple_items[0]),
+            mp_obj_get_int(tuple_items[1]),
+            value);
+
+        // add all 4 surrounding points to the list to check
```

Comment:
May be my ignorance on displayio but does something in here check the bitmap size? Like if I am on row 0 and can it check row -1? Maybe in the case a rectangle is only half on the screen and clipped so there isn't a bound in one direction.

Reply:
I dont think there is anything inherent to displayio that will prevent that from breaking. 

Nice catch thank you. Ill work on a fix for that and add another region near the edge in the test script to ensure it works.

Reply:
It seems that this lack of bounds checking was at least one of the culprits behind the MemoryErrors. After adding the bounds validation to ensure it won't check points that are outside of the bitmap I am able to fill a bitmap the entire size of the Pyportal screen 320x240 (albeit somewhat slowly), without a MemoryError being raised. 

Without the bounds check it seems like it became an infinite loop continually adding more and more points that were outside the bitmap.

I am still curious if the way that it is creating arrays or tuples could be made more efficient by re-using one or both somehow but adding the bounds check does at least resolve the worst case scenario.

Commit `6f78306` adds the checks to ignore points outside of the bitmap using this newer version I tested by modifying the example script posted above with a new rectangle section on each of the 4 sides:
```
# left edge
bitmaptools.draw_line(bitmap, 0, 60, 40, 60, 1)
bitmaptools.draw_line(bitmap, 0, 100, 40, 100, 1)
bitmaptools.draw_line(bitmap, 40, 60, 40, 100, 1)
bitmaptools.boundary_fill(bitmap,5, 65, 2, 0)

# top edge
bitmaptools.draw_line(bitmap, 200, 0, 200, 60, 1)
bitmaptools.draw_line(bitmap, 200, 60, 240, 60, 1)
bitmaptools.draw_line(bitmap, 240, 0, 240, 60, 1)
bitmaptools.boundary_fill(bitmap,202, 4, 1, 0)

# right edge
bitmaptools.draw_line(bitmap, 319, 100, 300, 100, 1)
bitmaptools.draw_line(bitmap, 319, 130, 300, 130, 1)
bitmaptools.draw_line(bitmap, 300, 100, 300, 130, 1)
bitmaptools.boundary_fill(bitmap,315, 110, 2, 0)

# bottom edge
bitmaptools.draw_line(bitmap, 210, 239, 210, 200, 1)
bitmaptools.draw_line(bitmap, 190, 239, 190, 200, 1)
bitmaptools.draw_line(bitmap, 210, 200, 190, 200, 1)
bitmaptools.boundary_fill(bitmap,200, 238, 1, 0)
```
And it now looks like this when it runs:
![image](https://user-images.githubusercontent.com/2406189/129496956-5169b5f4-a16b-4893-ba5d-ef535be98fe2.png)

Reply:
I cannot comment on the the performance questions, but the code to stop it from going outside the bitmap looks good to me. Glad that stopped the memory error.